tools/samtools.py

```diff
@@ -182,6 +182,59 @@ def removeDoublyMappedReads(self, inBam, outBam):
         opts = ['-b', '-F' '1028', '-f', '2', '-@', '3']
         self.view(opts, inBam, outBam)
 
+    def filter_to_mapped_reads(self, inBam, outBam, allow_unmapped=True, min_mapping_qual=None, remove_singletons=True):
+        '''
+            This function writes a bam file filtered to include properly aligned reads.
+            If allow_unmapped=True, fully-unmapped pairs or unmapped single-end reads are also 
```

Comment:
I think the docstring could use some explicit description of the other two optional parameters, and how they all interact:
- `min_mapping_qual` presumably tosses an entire alignment (paired or unpaired) -- or does it only toss one read alignment at a time or both mate pairs (if either of their mapqs are too low)? Put another way, does the `remove_singletons` act post-min-mapping-qual or pre-min-mapping-qual?
- `remove_singletons` -- does this toss unpaired unmapped reads, or just paired reads where only one mapped? Does it toss "improperly paired" reads (both reads in a mate-pair map, but perhaps point the wrong way on the same ref sequence, or perhaps map to different ref sequences)

Reply:
Hm.. so in my mind, maybe a simple logical flow that makes it conceptually a little easier to think about might be something like a series of optional params like:
1. `min_mapping_qual` - remove any alignment below this mapping qual
2. `remove_improper_pairs` - remove any alignments for any read from a paired-end sequencing run where both mates are not aligned
3. `remove_unmapped` - toss any reads that are not aligned
And processed in the above order: first 1, then 2, then 3.

Possibly even add a "remove secondary alignments" between 1 and 2 (ie, toss all the alignments that are not the primary mapping, for the aligners that do that kind of thing).

The problem with this actually.... is that I don't know if we can do this in a streaming/iterator way (w/o accumulating reads in memory) because you can't know if a read has its mate also aligned w/o seeing that mate, and aligners always emit their stuff in coordinate sort order, not query name sort order, so the mate might be far downstream (if it exists at all). We'd need to cache entries in memory in a buffer that grows on the order of the number of reads that align as singles/unmated (and then we'd have to re-sort the output again to comply with the stated sort order).

Maybe this problem becomes simpler actually if we remove the `min_mapping_qual` filter from this function. Because then all the questions become immediately answerable by looking at the boolean SAM flags alone (because the `min_mapping_qual` filtering potentially makes the boolean flags no longer correct). If we get rid of the min-mapping functionality and just have `remove_improper_pairs` and `remove_unmapped` then we can go soley on the combination of bit flags.

Reply:
Thanks for taking such a close look at this and thinking through the process. The part we can't do with samtools filtering without a split+merge is having two code paths for single-end and paired end reads, where single-end reads are allowed through but paired-end singleton reads are removed. Maybe that should be the focus here, and the quality filtering can be added via a samtools call immediately upstream. Then we keep only mapped single-end reads and properly mapped pairs?